File: backend/app/schemas/oa/workflow.py
"""
OA工作流Schema定义

包含工作流定义、审批节点、审批实例、审批任务、流程变量的Pydantic Schema
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID

from pydantic import BaseModel, Field


# ============================================================================
# WorkflowDefinition Schemas
# ============================================================================

class WorkflowDefinitionCreate(BaseModel):
    """创建工作流定义"""
    name: str = Field(..., min_length=1, max_length=100, description="工作流名称")
    code: str = Field(..., min_length=1, max_length=50, description="工作流编码")
    description: Optional[str] = Field(None, description="工作流描述")
    business_type: Optional[str] = Field(None, max_length=50, description="关联业务类型")
    config: Optional[Dict[str, Any]] = Field(None, description="工作流配置")
    form_config: Optional[Dict[str, Any]] = Field(None, description="表单配置")
    allow_withdraw: bool = Field(True, description="允许撤回")
    allow_transfer: bool = Field(True, description="允许转交")
    allow_cc: bool = Field(True, description="允许抄送")


class WorkflowDefinitionUpdate(BaseModel):
    """更新工作流定义"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="工作流名称")
    description: Optional[str] = Field(None, description="工作流描述")
    config: Optional[Dict[str, Any]] = Field(None, description="工作流配置")
    form_config: Optional[Dict[str, Any]] = Field(None, description="表单配置")
    is_active: Optional[bool] = Field(None, description="是否启用")
    allow_withdraw: Optional[bool] = Field(None, description="允许撤回")
    allow_transfer: Optional[bool] = Field(None, description="允许转交")
    allow_cc: Optional[bool] = Field(None, description="允许抄送")


class WorkflowDefinitionOut(BaseModel):
    """工作流定义输出"""
    id: UUID
    name: str
    code: str
    description: Optional[str] = None
    version: int
    is_active: bool
    business_type: Optional[str] = None
    config: Optional[Dict[str, Any]] = None
    form_config: Optional[Dict[str, Any]] = None
    allow_withdraw: bool
    allow_transfer: bool
    allow_cc: bool
    published_at: Optional[datetime] = None
    published_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class WorkflowDefinitionListOut(BaseModel):
    """工作流定义列表项"""
    id: UUID
    name: str
    code: str
    version: int
    is_active: bool
    business_type: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# ============================================================================
# WorkflowNode Schemas
# ============================================================================

class WorkflowNodeCreate(BaseModel):
    """创建审批节点"""
    name: str = Field(..., min_length=1, max_length=100, description="节点名称")
    code: Optional[str] = Field(None, max_length=50, description="节点编码")
    node_type: str = Field(..., description="节点类型: START/END/APPROVAL/CONDITION/CC/AUTO")
    order_index: int = Field(0, description="顺序索引")
    config: Optional[Dict[str, Any]] = Field(None, description="节点配置")
    approver_rule: Optional[Dict[str, Any]] = Field(None, description="审批人规则")
    timeout_hours: Optional[int] = Field(None, description="超时时限(小时)")
    timeout_action: Optional[str] = Field(None, description="超时动作: SKIP/AUTO_APPROVE/NOTIFY")
    condition_expression: Optional[str] = Field(None, description="条件表达式")
    group_id: Optional[UUID] = Field(None, description="节点分组ID")
    is_required: bool = Field(True, description="是否必须审批")


class WorkflowNodeUpdate(BaseModel):
    """更新审批节点"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="节点名称")
    code: Optional[str] = Field(None, max_length=50, description="节点编码")
    node_type: Optional[str] = Field(None, description="节点类型")
    order_index: Optional[int] = Field(None, description="顺序索引")
    config: Optional[Dict[str, Any]] = Field(None, description="节点配置")
    approver_rule: Optional[Dict[str, Any]] = Field(None, description="审批人规则")
    timeout_hours: Optional[int] = Field(None, description="超时时限(小时)")
    timeout_action: Optional[str] = Field(None, description="超时动作")
    condition_expression: Optional[str] = Field(None, description="条件表达式")
    is_required: Optional[bool] = Field(None, description="是否必须审批")


class WorkflowNodeOut(BaseModel):
    """审批节点输出"""
    id: UUID
    definition_id: UUID
    name: str
    code: Optional[str] = None
    node_type: str
    order_index: int
    config: Optional[Dict[str, Any]] = None
    approver_rule: Optional[Dict[str, Any]] = None
    timeout_hours: Optional[int] = None
    timeout_action: Optional[str] = None
    condition_expression: Optional[str] = None
    group_id: Optional[UUID] = None
    is_required: bool
    created_at: datetime

    class Config:
        from_attributes = True


# ============================================================================
# WorkflowInstance Schemas
# ============================================================================

class WorkflowInstanceCreate(BaseModel):
    """创建审批实例"""
    definition_id: UUID = Field(..., description="工作流定义ID")
    business_type: str = Field(..., max_length=50, description="业务类型")
    business_id: UUID = Field(..., description="业务数据ID")
    title: str = Field(..., max_length=200, description="审批标题")
    summary: Optional[str] = Field(None, description="审批摘要")
    form_data: Optional[Dict[str, Any]] = Field(None, description="表单数据")
    cc_list: Optional[List[UUID]] = Field(None, description="抄送用户ID列表")


class WorkflowInstanceOut(BaseModel):
    """审批实例输出"""
    id: UUID
    definition_id: UUID
    definition: Optional[WorkflowDefinitionOut] = None
    business_type: str
    business_id: UUID
    initiator_id: UUID
    initiator_name: Optional[str] = None
    status: str
    current_node_id: Optional[UUID] = None
    completed_node_ids: Optional[List[UUID]] = None
    approval_summary: Optional[Dict[str, Any]] = None
    cc_list: Optional[List[UUID]] = None
    form_data: Optional[Dict[str, Any]] = None
    title: str
    summary: Optional[str] = None
    submitted_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    cancel_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class WorkflowInstanceListOut(BaseModel):
    """审批实例列表项"""
    id: UUID
    definition_id: UUID
    definition_name: Optional[str] = None
    business_type: str
    business_id: UUID
    initiator_id: UUID
    initiator_name: Optional[str] = None
    status: str
    title: str
    submitted_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class WorkflowInstanceCancel(BaseModel):
    """撤回审批实例"""
    reason: Optional[str] = Field(None, description="撤回原因")


# ============================================================================
# WorkflowTask Schemas
# ============================================================================

class WorkflowTaskOut(BaseModel):
    """审批任务输出"""
    id: UUID
    instance_id: UUID
    node_id: UUID
    node_name: Optional[str] = None
    task_type: str
    status: str
    assignee_id: Optional[UUID] = None
    assignee_name: Optional[str] = None
    assignee_type: Optional[str] = None
    original_assignee_id: Optional[UUID] = None
    assigned_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    deadline: Optional[datetime] = None
    is_overdue: bool
    overdue_reminders: int
    comment: Optional[str] = None
    action: Optional[str] = None
    order_index: int
    is_required: bool
    created_at: datetime

    class Config:
        from_attributes = True


class WorkflowTaskListOut(BaseModel):
    """审批任务列表项"""
    id: UUID
    instance_id: UUID
    node_id: UUID
    node_name: Optional[str] = None
    task_type: str
    status: str
    assignee_id: Optional[UUID] = None
    assignee_name: Optional[str] = None
    title: Optional[str] = None  # 来自instance
    business_type: Optional[str] = None
    assigned_at: Optional[datetime] = None
    deadline: Optional[datetime] = None
    is_overdue: bool

    class Config:
        from_attributes = True


class TaskAction(BaseModel):
    """审批任务操作"""
    action: str = Field(..., description="操作类型: APPROVE/REJECT/TRANSFER/DELEGATE")
    comment: Optional[str] = Field(None, description="审批意见")
    transfer_to: Optional[UUID] = Field(None, description="转交给谁")
    transfer_reason: Optional[str] = Field(None, description="转交原因")
    delegate_to: Optional[UUID] = Field(None, description="代理给谁")


class TaskTransfer(BaseModel):
    """转交任务"""
    transfer_to: UUID = Field(..., description="转交给谁")
    reason: Optional[str] = Field(None, description="转交原因")


class TaskDelegate(BaseModel):
    """代理任务"""
    delegate_to: UUID = Field(..., description="代理给谁")
    reason: Optional[str] = Field(None, description="代理原因")


# ============================================================================
# WorkflowCC Schemas
# ============================================================================

class WorkflowCCOut(BaseModel):
    """抄送记录输出"""
    id: UUID
    instance_id: UUID
    instance_title: Optional[str] = None
    business_type: Optional[str] = None
    cc_user_id: UUID
    cc_user_name: Optional[str] = None
    cc_at: datetime
    is_read: bool
    read_at: Optional[datetime] = None
    reason: Optional[str] = None

    class Config:
        from_attributes = True


class WorkflowCCMarkRead(BaseModel):
    """标记抄送已读"""
    pass


# ============================================================================
# WorkflowVariable Schemas
# ============================================================================

class WorkflowVariableSchema(BaseModel):
    """流程变量Schema"""
    name: str = Field(..., max_length=100, description="变量名称")
    value: Optional[str] = Field(None, description="变量值")
    value_type: str = Field("STRING", description="类型: STRING/NUMBER/BOOLEAN/DATE/JSON/ARRAY")
    source: Optional[str] = Field(None, description="来源: SYSTEM/FORM/APPROVAL/CONDITION")


class WorkflowVariableCreate(WorkflowVariableSchema):
    """创建流程变量"""
    pass


class WorkflowVariableOut(WorkflowVariableSchema):
    """流程变量输出"""
    id: UUID
    instance_id: UUID
    node_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ============================================================================
# Query Schemas
# ============================================================================

class WorkflowInstanceQuery(BaseModel):
    """审批实例查询参数"""
    status: Optional[str] = Field(None, description="状态过滤")
    business_type: Optional[str] = Field(None, description="业务类型过滤")
    keyword: Optional[str] = Field(None, description="关键词搜索")
    start_date: Optional[datetime] = Field(None, description="开始日期")
    end_date: Optional[datetime] = Field(None, description="结束日期")


class WorkflowTaskQuery(BaseModel):
    """审批任务查询参数"""
    status: Optional[str] = Field(None, description="状态过滤")
    task_type: Optional[str] = Field(None, description="任务类型过滤")
    is_overdue: Optional[bool] = Field(None, description="是否超时")


class WorkflowCCQuery(BaseModel):
    """抄送记录查询参数"""
    is_read: Optional[bool] = Field(None, description="是否已读")
    business_type: Optional[str] = Field(None, description="业务类型过滤")

File: frontend/src/types/workflow.ts
/**
 * 工作流模块类型定义
 * 与后端 Pydantic Schema 及 SQLAlchemy Model 对齐
 */

// ============================================================
// 枚举类型
// ============================================================

/** 工作流实例状态 */
export enum InstanceStatus {
  PENDING = 'PENDING',       // 待审批（已提交，等待第一个审批人）
  APPROVING = 'APPROVING',  // 审批中
  APPROVED = 'APPROVED',     // 已通过
  REJECTED = 'REJECTED',     // 已拒绝
  CANCELLED = 'CANCELLED',   // 已撤回
  EXPIRED = 'EXPIRED',       // 已超时
}

/** 工作流节点类型 */
export enum NodeType {
  START = 'START',       // 开始节点
  END = 'END',           // 结束节点
  APPROVAL = 'APPROVAL', // 审批节点
  CONDITION = 'CONDITION', // 条件节点
  CC = 'CC',             // 抄送节点
}

/** 审批动作 */
export enum ApproveAction {
  APPROVE = 'APPROVE',   // 通过
  REJECT = 'REJECT',      // 拒绝
  TRANSFER = 'TRANSFER',  // 转交
  DELEGATE = 'DELEGATE',  // 转派
  URGE = 'URGE',          // 催办
}

/** 超时处理方式 */
export enum TimeoutAction {
  NOTIFY = 'NOTIFY',         // 仅提醒
  AUTO_APPROVE = 'AUTO_APPROVE', // 自动通过
  AUTO_REJECT = 'AUTO_REJECT',   // 自动拒绝
  ESCALATE = 'ESCALATE',     // 升级处理
}

/** 审批人规则类型 */
export enum ApproverRuleType {
  USER = 'USER',             // 指定用户
  ROLE = 'ROLE',             // 指定角色
  DEPARTMENT_LEADER = 'DEPARTMENT_LEADER', // 部门主管
  MULTI_LEVEL_LEADER = 'MULTI_LEVEL_LEADER', // 多级主管
  FORM_FIELD = 'FORM_FIELD', // 表单字段
  SCRIPT = 'SCRIPT',         // 脚本计算
  CUSTOM = 'CUSTOM',         // 自定义
}

// ============================================================
// 条件表达式类型
// ============================================================

/** 简单条件（三段式） */
export interface SimpleCondition {
  field: string;      // 字段路径，如 "form.amount"
  operator: '==' | '!=' | '>' | '>=' | '<' | '<=' | 'in' | 'not_in' | 'contains';
  value: string | number | boolean | string[] | number[];
}

/** 条件表达式（支持嵌套） */
export interface ConditionExpression {
  op: 'AND' | 'OR' | 'NOT';
  conditions?: ConditionExpression[];
  condition?: SimpleCondition;  // NOT 操作符使用
}

// ============================================================
// 审批人规则
// ============================================================

export interface ApproverRule {
  type: ApproverRuleType;
  user_ids?: string[];
  role_ids?: string[];
  department_ids?: string[];
  field_name?: string;        // FORM_FIELD 类型时使用的表单字段名
  script?: string;            // SCRIPT 类型时的脚本内容
}

// ============================================================
// 工作流节点
// ============================================================

export interface WorkflowNode {
  id: string;
  name: string;
  code: string;
  node_type: NodeType;
  order_index: number;
  approver_rule?: ApproverRule;
  config?: Record<string, any>;
  timeout_hours?: number;
  timeout_action?: TimeoutAction;
  condition_expression?: ConditionExpression;
  is_current?: boolean;
  is_completed?: boolean;
  is_pending?: boolean;
  operator_name?: string;
  completed_at?: string;
  tasks?: WorkflowTask[];
}

/** 工作流定义状态 */
export enum DefinitionStatus {
  DRAFT = 'draft',       // 草稿状态（不可发起审批）
  PUBLISHED = 'published', // 已发布（可发起审批）
  DISABLED = 'disabled',  // 已禁用（不可发起审批）
}

// ============================================================
// 工作流定义
// ============================================================

export interface WorkflowDefinition {
  id: string;
  name: string;
  code: string;
  description?: string;
  category?: string;
  business_type?: string;
  version: number;
  status: DefinitionStatus;  // 工作流定义状态: draft/published/disabled
  is_active: boolean;        // 是否启用（功能开关）
  form_config?: FormConfig;
  allow_withdraw: boolean;
  allow_transfer: boolean;
  allow_cc: boolean;
  allow_urge?: boolean;
  allow_all_initiator?: boolean;
  allowed_roles?: string[];
  allowed_departments?: string[];
  nodes?: WorkflowNode[];
  instance_count?: number;
  created_at?: string;
  updated_at?: string;
  published_at?: string;
  published_by?: string;
}

/** 表单字段配置 */
export interface FormField {
  name: string;
  label: string;
  type: 'text' | 'textarea' | 'number' | 'date' | 'datetime' | 'select' | 'checkbox' | 'radio' | 'file';
  required?: boolean;
  options?: { label: string; value: string }[];
  placeholder?: string;
  default_value?: any;
}

/** 表单配置 */
export interface FormConfig {
  fields: FormField[];
  layout?: 'single' | 'double';
}

// ============================================================
// 工作流实例
// ============================================================

export interface WorkflowInstance {
  id: string;
  instance_id: string;
  workflow_id: string;
  workflow_name?: string;
  workflow_code?: string;
  business_type?: string;
  business_id?: string;
  title: string;
  description?: string;
  form_data?: Record<string, any>;
  form_fields?: FormField[];
  status: InstanceStatus;
  result?: string;
  current_node_id?: string;
  current_node_name?: string;
  can_withdraw?: boolean;
  can_urge?: boolean;
  urge_count?: number;
  last_urge_at?: string;
  is_overdue?: boolean;
  initiator_id: string;
  initiator_name?: string;
  created_at?: string;
  updated_at?: string;
  completed_at?: string;
  nodes?: WorkflowNode[];
  histories?: WorkflowHistory[];
  attachments?: WorkflowAttachment[];
  cc_list?: WorkflowCC[];
}

// ============================================================
// 审批任务
// ============================================================

export interface WorkflowTask {
  id: string;
  instance_id: string;
  node_id: string;
  node_name?: string;
  workflow_name?: string;
  title?: string;
  assignee_id: string;
  assignee_name?: string;
  action?: ApproveAction;
  comment?: string;
  status: 'PENDING' | 'COMPLETED' | 'TRANSFERRED' | 'DELEGATED';
  is_read?: boolean;
  is_overdue?: boolean;
  due_at?: string;
  processed_at?: string;
  created_at?: string;
  updated_at?: string;
}

// ============================================================
// 审批历史
// ============================================================

export interface WorkflowHistory {
  id: string;
  action: ApproveAction;
  node_name?: string;
  operator_id: string;
  operator_name?: string;
  comment?: string;
  attachment_count?: number;
  created_at?: string;
}

// ============================================================
// 抄送
// ============================================================

export interface WorkflowCC {
  id: string;
  instance_id: string;
  user_id: string;
  user_name?: string;
  is_read?: boolean;
  read_at?: string;
  created_at?: string;
}

// ============================================================
// 附件
// ============================================================

export interface WorkflowAttachment {
  id: string;
  filename: string;
  size: number;
  url: string;
  uploaded_by?: string;
  uploaded_at?: string;
}

// ============================================================
// 统计数据
// ============================================================

/** 统计摘要 */
export interface WorkflowStatisticsSummary {
  total: number;
  approved: number;
  rejected: number;
  cancelled: number;
  in_progress: number;
  approve_rate: number;
  avg_duration_hours: number;
}

/** 按类型统计 */
export interface WorkflowStatisticsByType {
  business_type: string;
  total: number;
  approved: number;
  rejected: number;
  approve_rate: number;
}

/** 完整统计数据 */
export interface WorkflowStatistics {
  summary: WorkflowStatisticsSummary;
  by_type: WorkflowStatisticsByType[];
}

// ============================================================
// 请求参数类型
// ============================================================

/** 发起审批请求 */
export interface StartInstanceRequest {
  workflow_id: string;
  business_type: string;
  business_id: string;
  title: string;
  context?: Record<string, any>;
  cc_list?: string[];
  summary?: string;
}

/** 催办请求 */
export interface UrgeRequest {
  message?: string;
}

/** 审批操作请求 */
export interface ApproveRequest {
  comment?: string;
}

/** 拒绝请求 */
export interface RejectRequest {
  comment: string;
}

/** 转交请求 */
export interface TransferRequest {
  target_user_id: string;
  comment?: string;
}

/** 转派请求 */
export interface DelegateRequest {
  delegate_to: string;
  comment?: string;
}

/** 工作流定义查询参数 */
export interface WorkflowDefinitionQuery {
  page?: number;
  page_size?: number;
  name?: string;
  status?: 'active' | 'inactive';
}

/** 工作流实例查询参数 */
export interface WorkflowInstanceQuery {
  page?: number;
  page_size?: number;
  status?: InstanceStatus;
  business_type?: string;
}

/** 工作流统计查询参数 */
export interface WorkflowStatisticsQuery {
  start_date?: string;   // YYYY-MM-DD
  end_date?: string;     // YYYY-MM-DD
  business_type?: string;
}

/** 审批任务查询参数 */
export interface WorkflowTaskQuery {
  page?: number;
  page_size?: number;
  status?: string;
}

// ============================================================
// 分页响应
// ============================================================

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

// ============================================================
// 后端响应包装
// ============================================================

export interface ApiResponse<T> {
  code: number;
  message?: string;
  data: T;
}

export interface ApiPageResponse<T> {
  code: number;
  message?: string;
  data: {
    items: T[];
    total: number;
    page: number;
    page_size: number;
  };
}
